adal/src/main/java/com/microsoft/aad/adal/BrokerProxy.java

```diff
@@ -558,34 +567,154 @@ private boolean hasSupportToAddUserThroughBroker(final String brokerPackageName)
+        if (!mBrokerTag.equals(tag) &&
+                !AuthenticationConstants.Broker.AZURE_AUTHENTICATOR_APP_SIGNATURE.equals(tag)) {
+            throw new AuthenticationException(ADALError.BROKER_APP_VERIFICATION_FAILED);
+        }
+
+        return true;
+    }
+
+    private List<X509Certificate> readCertDataForBrokerApp(final String brokerPackageName)
+            throws NameNotFoundException, AuthenticationException, IOException,
+            GeneralSecurityException {
+        final PackageInfo packageInfo = mContext.getPackageManager().getPackageInfo(brokerPackageName,
+                PackageManager.GET_SIGNATURES);
+        if (packageInfo == null) {
+            throw new AuthenticationException(ADALError.APP_PACKAGE_NAME_NOT_FOUND,
+                    "No broker package existed.");
+        }
+
+        if (packageInfo.signatures == null || packageInfo.signatures.length == 0) {
+            throw new AuthenticationException(ADALError.BROKER_APP_VERIFICATION_FAILED,
+                    "No signature associated with the broker package.");
+        }
+
+        final List<X509Certificate> certificates = new ArrayList<>(packageInfo.signatures.length);
+        for (final Signature signature : packageInfo.signatures) {
+            final byte[] rawCert = signature.toByteArray();
+            final InputStream certStream = new ByteArrayInputStream(rawCert);
+
+            final CertificateFactory certificateFactory;
+            final X509Certificate x509Certificate;
+            try {
+                certificateFactory = CertificateFactory.getInstance("X509");
+                x509Certificate = (X509Certificate) certificateFactory.generateCertificate(
+                        certStream);
+                certificates.add(x509Certificate);
+            } catch (final CertificateException e) {
+                throw new AuthenticationException(ADALError.BROKER_APP_VERIFICATION_FAILED);
+            }
+        }
+
+        return certificates;
+    }
+
+    private X509Certificate getSignerCertAndVerifyCertificateChain(
+            final List<X509Certificate> certificates)
+            throws GeneralSecurityException, AuthenticationException {
+        // create certificate chain, find the self signed cert first and chain all the way back
+        // to the signer cert. Also perform certificate signing validation when chaining them back.
+        X509Certificate issuerCert = getSelfSignedCert(certificates);
+        X509Certificate intermediateCert;
+        int count = 1;
+        while (true) {
+            intermediateCert = findCert(issuerCert.getSubjectDN(),
+                    certificates);
+            if (intermediateCert == null) {
+                throw new AuthenticationException(ADALError.BROKER_APP_VERIFICATION_FAILED);
+            }
+
+            //verify certificate is signed with the public key of the issuer certificate
+            intermediateCert.verify(issuerCert.getPublicKey());
```

Comment:
[https://developer.android.com/reference/java/security/cert/X509Certificate.html](https://developer.android.com/reference/java/security/cert/X509Certificate.html)
says that verify was added in Android N SDK, how is this working?